rust/lance/src/dataset/write/commit.rs

```diff
@@ -0,0 +1,339 @@
+    /// passed, the storage format will be inferred from the data files.
+    ///
+    /// All data files must use the same storage format as the existing dataset.
+    /// If a different format is passed, an error will be returned.
+    pub fn with_storage_format(mut self, storage_format: LanceFileVersion) -> Self {
+        self.storage_format = Some(storage_format);
+        self
+    }
+
+    /// Pass an object store to use.
+    pub fn with_object_store(mut self, object_store: Arc<ObjectStore>) -> Self {
+        self.object_store = Some(object_store);
+        self
+    }
+
+    /// Pass a commit handler to use for the dataset.
+    pub fn with_commit_handler(mut self, commit_handler: Arc<dyn CommitHandler>) -> Self {
+        self.commit_handler = Some(commit_handler);
+        self
+    }
+
+    /// Pass store parameters to use for the dataset.
+    ///
+    /// If an object store is passed, these parameters will be ignored.
+    pub fn with_store_params(mut self, store_params: ObjectStoreParams) -> Self {
+        self.store_params = Some(store_params);
+        self
+    }
+
+    /// Pass an object store registry to use.
+    ///
+    /// If an object store is passed, this registry will be ignored.
+    pub fn with_object_store_registry(
+        mut self,
+        object_store_registry: Arc<ObjectStoreRegistry>,
+    ) -> Self {
+        self.object_store_registry = object_store_registry;
+        self
+    }
+
+    /// Pass a session to use for the dataset.
+    ///
+    /// If a session is not passed, but a dataset is used as the destination,
+    /// then the dataset's session will be used.
+    ///
+    /// By passing a session or re-using a dataset, you can re-use the
+    /// file metadata and index caches, which can significantly improve
+    /// performance.
+    pub fn with_session(mut self, session: Arc<Session>) -> Self {
+        self.session = Some(session);
+        self
+    }
+
+    ///  If set to true, and this is a new dataset, uses the new v2 manifest
+    ///  paths. These allow constant-time lookups for the latest manifest on object storage.
+    ///  This parameter has no effect on existing datasets. To migrate an existing
+    ///  dataset, use the [`Dataset::migrate_manifest_paths_v2`] method. **Default is False.**
+    ///
+    /// <div class="warning">
+    ///  WARNING: turning this on will make the dataset unreadable for older
+    ///  versions of Lance (prior to 0.17.0).
+    /// </div>
+    pub fn enable_v2_manifest_paths(mut self, enable: bool) -> Self {
+        self.enable_v2_manifest_paths = enable;
+        self
+    }
+
+    /// Commit a version that is not part of the mainline history.
+    ///
+    /// This commit will never show up in the dataset's history.
+    ///
+    /// This can be used to stage changes or to handle "secondary" datasets
+    /// whose lineage is tracked elsewhere.
+    pub fn with_detached(mut self, detached: bool) -> Self {
```

Comment:
interesting, sounds like git

Reply:
anybody is using this feature or why this feature is asked?

Reply:
It's kind of like git. Though there only one branch, and everything else is detached.

Weston added this feature to support "balanced storage" for large blob columns. Basically we used this to create a separate hidden dataset that stores the blob data, and by doing this we could compact that data at a different rate than other columns.

Reply:
Interesting, may be with 2.1 we don't need to have this kind of detached dataset for blob?

Reply:
Actually it was designed with 2.1 in mind. The ideal rows per file differs so much between small and wide columns that they need to essentially be in different datasets to offer good enough OLAP and random access performance. We wrap it up quite seamlessly though, so for the most part it feels like just another column.

Reply:
Yeah, for example, imagine a column contains 1MB images.  1M rows would mean 1TB for just that one column.  1TB files are difficult to work with in cloud storage (operations take a long time, sometimes they fail) and running compaction (e.g. on two 500K files) is very slow.

With balanced storage we have two datasets.  The primary dataset has all the smaller columns.  The blob dataset has all the larger columns.  Each dataset targets a reasonable size for files (e.g. 64GB).  The primary dataset has fewer fragments but each fragment has at least 1M rows (important for small data types).  The secondary dataset has more fragments but the data types are so large we don't care about the # rows per files as much.